src/audio/google/google_rtc_audio_processing.c

```diff
@@ -676,6 +676,31 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
 		comp_warn(dev, "Too many mic channels: %d, truncating to %d",
 			  mic_chan, CHAN_MAX);
 
+#ifdef CONFIG_IPC_MAJOR_4
+	/* Bug workaround: IPC4 doesn't correctly initialize metadata
+	 * on (at least) our reference source.  Use our
+	 * kconfig-derived value instead, but yell about it.  Note
+	 * that we MUST correct this at the source level and not just
+	 * re-interpret it ourselves, otherwise other elements of the
+	 * pipeline get confused.
+	 */
+	if (ref_chan != CHAN_MAX) {
+		comp_err(dev, "Incorrect ref channels %d, setting %d",
+			 ref_chan, CHAN_MAX);
+		ref_chan = CHAN_MAX;
+		source_set_channels(sources[cd->aec_reference_source], ref_chan);
+	}
+
+	/* IPC4 also fails to call this function (which isn't merely a
+	 * field set, it recomputes other values that are otherwise
+	 * left as garbage) on our input stream at setup time, leading
+	 * to confusing failures with buffering behavior.
+	 */
+	source_set_alignment_constants(sources[0], source_get_frame_bytes(sources[0]), 1);
+	source_set_alignment_constants(sources[1], source_get_frame_bytes(sources[1]), 1);
+	sink_set_alignment_constants(sinks[0], sink_get_frame_bytes(sinks[0]), 1);
```

Comment:
@andrula-song  Can you help to check this alignment setting ? why this module need this alignment ?  We have setting in framework with "	audio_stream_init_alignment_constants(1, 1, &buffer->stream);".   Each module set the alignment to its sink buffer, so we don't need to set its source buffer here.

Reply:
1.it makes sense to set the frame_align_req as 1 and byte_align as frame size(instead of 1 byte) for component rtc.
2. seems this is a sink/source API module, so use source_set_alignment_constants/sink_set_alignment_constants is right here.

Reply:
actually, I want to know more detail failure info of the comment. I am not sure would this explain can help: take stereo 24/24 stream as example, the frame size is 2*3 = 6 byte; with the setting frame_alig_req = 1 and byte_align = 6, we will get the align_shift_idx as 3(for 8 bytes). Assume there are 47 frames data need to process, if we use audio_stream_avail_frames_aligned to get the available frames, you will get (47 * 6) >> 3 = 35 frames, and 47-35 = 12 frames left to process in next time.

but from your code, you're using source_get_data_frames_available &  sink_get_free_size to calculate the available  and free size of the buffer, so the _aligned setting is never used in your case, you can just remove those 3 lines, it shouldn't affect anything.

Reply:
The comment tries to explain, but it breaks audio if I remove those lines, something starts overflowing the ring buffers and the output is distorted.  See bug #8639 

I don't want them here either, and I thought they weren't required also.  That turns out to have been wrong.

Reply:
NP, we can add TODO to fix it by our team later